File: javascript_version/mixed_recall/binary_search/binary_search.js
// binary search
// Write a function, binarySearch, that takes in a sorted array of numbers and a target. 
// The function should return the index where the target can be found within the array. 
// If the target is not found in the array, then return -1.

// You may assume that the input array contains unique numbers sorted in increasing order.

// Your function must implement the binary search algorithm.
// https://en.wikipedia.org/wiki/Binary_search_algorithm


const binarySearch = (numbers, target) => {
    // todo
    let first = 0;
    let last = numbers.length - 1;
    let mid = Math.floor((first + last) / 2);

    while (first <= last) {

        if (numbers[mid] > target) {
            last = mid - 1;
        }
        else if (numbers[mid] === target) {
            return mid;
        }
        else if (numbers[mid] < target) {
            first = mid + 1;
        }
        mid = Math.floor((first + last) / 2);
    }

    return -1;
};


const recursiveBinarySearch = (numbers, target) => {
    if (numbers.length === 0) { return -1; }
    let mid = Math.floor((numbers.length - 1) / 2);
    if (numbers[mid] === target) { return mid; }
    if (numbers[mid] > target) { return recursiveBinarySearch(numbers.slice(0, mid), target); }
    if (numbers[mid] < target) {
        let result = recursiveBinarySearch(numbers.slice(mid + 1), target);
        if (result === -1) { return -1; }
        return mid + 1 + result;
    }


};
console.log(binarySearch([0, 1, 2, 3, 4, 5, 6, 7, 8], 6)); // -> 6
console.log(binarySearch([0, 6, 8, 12, 16, 19, 20, 24, 28], 27)); // -> -1
console.log(binarySearch([0, 6, 8, 12, 16, 19, 20, 28], 8)); // -> 2
console.log(binarySearch([0, 6, 8, 12, 16, 19, 20, 24, 28], 28)); // -> 8
console.log(binarySearch([7, 9], 7)); // -> 0
console.log(binarySearch([7, 9], 9)); // -> 1
console.log(binarySearch([7, 9], 12)); // -> -1
console.log(binarySearch([7], 7)); // -> 0
console.log(binarySearch([], 7)); // -> -1


console.log("RBS");
console.log(recursiveBinarySearch([0, 1, 2, 3, 4, 5, 6, 7, 8], 6)); // -> 6
console.log(recursiveBinarySearch([0, 6, 8, 12, 16, 19, 20, 24, 28], 27)); // -> -1
console.log(recursiveBinarySearch([0, 6, 8, 12, 16, 19, 20, 28], 8)); // -> 2
console.log(recursiveBinarySearch([0, 6, 8, 12, 16, 19, 20, 24, 28], 28)); // -> 8
console.log(recursiveBinarySearch([7, 9], 7)); // -> 0
console.log(recursiveBinarySearch([7, 9], 9)); // -> 1
console.log(recursiveBinarySearch([7, 9], 12)); // -> -1
console.log(recursiveBinarySearch([7], 7)); // -> 0
console.log(recursiveBinarySearch([], 7)); // -> -1

File: cpp_version/mixed_recall/binary_search/binary_search.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// binary search
// Write a function, binarySearch, that takes in a sorted array of numbers and a target.
// The function should return the index where the target can be found within the array.
// If the target is not found in the array, then return -1.

// You may assume that the input array contains unique numbers sorted in increasing order.

// Your function must implement the binary search algorithm.
// https://en.wikipedia.org/wiki/Binary_search_algorithm
int rbs(std::vector<int>, int);
int _rbs(std::vector<int>, int, int, int);
int binarySearch(std::vector<int>, int);
int recursiveBinarySearch(std::vector<int>, int);

int binarySearch(std::vector<int> numbers, int target) {
    // TODO

    return -1;
}

// recusive binary search old method
int rbs(std::vector<int> numbers, int target) {
    if (numbers.size() == 0) {
        return -1;
    }
    return _rbs(numbers, target, 0, numbers.size() - 1);
}

int _rbs(std::vector<int> numbers, int target, int first, int last) {



    return -1;
}

// recursive binary search 2 params
int recursiveBinarySearch(std::vector<int> numbers, int target) {
    

    return -1;
}

int main() {
    std::cout << binarySearch(std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7, 8}, 6) << std::endl;        // -> 6
    std::cout << binarySearch(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 24, 28}, 27) << std::endl; // -> -1
    std::cout << binarySearch(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 28}, 8) << std::endl;      // -> 2
    std::cout << binarySearch(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 24, 28}, 28) << std::endl; // -> 8
    std::cout << binarySearch(std::vector<int>{7, 9}, 7) << std::endl;                             // -> 0
    std::cout << binarySearch(std::vector<int>{7, 9}, 9) << std::endl;                             // -> 1
    std::cout << binarySearch(std::vector<int>{7, 9}, 12) << std::endl;                            // -> -1
    std::cout << binarySearch(std::vector<int>{7}, 7) << std::endl;                                // -> 0
    std::cout << binarySearch(std::vector<int>{}, 7) << std::endl;                                 // -> -1

    std::cout << "=============== RBS ====================" << std::endl;
    std::cout << rbs(std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7, 8}, 6) << std::endl;        // -> 6
    std::cout << rbs(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 24, 28}, 27) << std::endl; // -> -1
    std::cout << rbs(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 28}, 8) << std::endl;      // -> 2
    std::cout << rbs(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 24, 28}, 28) << std::endl; // -> 8
    std::cout << rbs(std::vector<int>{7, 9}, 7) << std::endl;                             // -> 0
    std::cout << rbs(std::vector<int>{7, 9}, 9) << std::endl;                             // -> 1
    std::cout << rbs(std::vector<int>{7, 9}, 12) << std::endl;                            // -> -1
    std::cout << rbs(std::vector<int>{7}, 7) << std::endl;                                // -> 0
    std::cout << rbs(std::vector<int>{}, 7) << std::endl;                                 // -> -1

    std::cout << "=============== RECURSIVE_BINARY_SEARCH ====================" << std::endl;

    std::cout << recursiveBinarySearch(std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7, 8}, 6) << std::endl;        // -> 6
    std::cout << recursiveBinarySearch(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 24, 28}, 27) << std::endl; // -> -1
    std::cout << recursiveBinarySearch(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 28}, 8) << std::endl;      // -> 2
    std::cout << recursiveBinarySearch(std::vector<int>{0, 6, 8, 12, 16, 19, 20, 24, 28}, 28) << std::endl; // -> 8
    std::cout << recursiveBinarySearch(std::vector<int>{7, 9}, 7) << std::endl;                             // -> 0
    std::cout << recursiveBinarySearch(std::vector<int>{7, 9}, 9) << std::endl;                             // -> 1
    std::cout << recursiveBinarySearch(std::vector<int>{7, 9}, 12) << std::endl;                            // -> -1
    std::cout << recursiveBinarySearch(std::vector<int>{7}, 7) << std::endl;                                // -> 0
    std::cout << recursiveBinarySearch(std::vector<int>{}, 7) << std::endl;                                 // -> -1

    return 0;
}